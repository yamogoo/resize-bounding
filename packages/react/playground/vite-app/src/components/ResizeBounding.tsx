import ResizeBoundingComponent, { type Props } from "react-resize-bounding";

const ResizeBounding = (props: Props) => {
  return (
    <ResizeBoundingComponent
      {...props}
      options={{
        knob: {
          show: true,
        },
      }}
    >
      {props.children}
    </ResizeBoundingComponent>
  );
};

export default ResizeBounding;
