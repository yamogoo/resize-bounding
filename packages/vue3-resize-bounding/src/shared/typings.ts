import fluentui from "@fluentui/merge-styles";

export namespace ResizeBounding {
  export interface Knob {
    show: boolean;
    normalHidden: boolean;
  }

  export enum SplitterPositions {
    CENTER = "central",
    EXTERNAL = "external",
    INTERNAL = "internal",
  }

  export type PanePosition = SplitterPositions | string;

  export interface Options {
    prefix: string;
    width: number;
    position: PanePosition;
    cursor: Partial<
      Record<"horizontal" | "vertical", CSSStyleDeclaration["cursor"]>
    >;
    knob: Partial<Knob>;
  }

  export type ResizeBoundingClassNames =
    | "container"
    | "pane"
    | "splitter"
    | "splitterContainer"
    | "knob";

  export interface IResizeBoundingClassNames {
    container: string;
    pane: string;
    splitter: string;
    splitterContainer: string;
    knob: string;
  }

  export interface IStyles extends Record<ResizeBoundingClassNames, IStyle> {}
  export type IStyle = fluentui.IStyle;
  export type IStyleSet = fluentui.IStyleSet;
}
