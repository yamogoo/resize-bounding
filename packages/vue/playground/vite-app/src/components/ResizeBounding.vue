<template>
  <ResizeBounding
    :options="{
      knob: {
        show: true,
      },
    }"
  >
    <slot></slot>
  </ResizeBounding>
</template>

<script setup lang="ts">
import ResizeBounding from "vue3-resize-bounding";
</script>
