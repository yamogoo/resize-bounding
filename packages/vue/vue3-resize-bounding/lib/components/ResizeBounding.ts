import { IStyles, Options, type PaneDirectionKey } from "../shared/typings";

export interface Props
  extends Partial<{
    width: number;
    height: number;
    minWidth: number;
    maxWidth: number;
    minHeight: number;
    maxHeight: number;
    disabled: boolean;
    directions: PaneDirectionKey;
    alwaysShowKnob: boolean;
    options: Partial<Options>;
    styles: Partial<IStyles>;
    updateWidth: (width: number) => void;
    updateHeight: (height: number) => void;
    dragStart: (dir: string) => void;
    dragMove: (dir: string) => void;
    dragEnd: (dir: string) => void;
    focus: (data: { state: boolean; direction: string }) => void;
  }> {}

export enum Emits {
  FOCUS = "focus",
  UPDATE_WIDTH = "update:width",
  UPDATE_HEIGHT = "update:height",
  DRAG_START = "drag:start",
  DRAG_MOVE = "drag:move",
  DRAG_END = "drag:end",
}
