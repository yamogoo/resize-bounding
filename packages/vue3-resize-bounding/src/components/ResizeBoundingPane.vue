<template>
  <div
    :class="[
      classNames.pane,
      `--${direction}`,
      options.addStateClasses && {
        normal: !isPressed && !isFocused,
        focused: isFocused,
        pressed: isPressed,
      },
    ]"
    data-testid="resize-bounding-pane"
    :style="[paneComputedStyle]"
  >
    <div
      ref="refPane"
      data-testid="resize-bounding-splitter"
      :class="[classNames.splitter]"
      :style="[splitterComputedStyle]"
    >
      <div
        v-if="
          (isFocused || !options?.knob?.normalHidden) && options?.knob?.show
        "
        data-testid="resize-bounding-splitter-container"
        :class="[classNames.splitterContainer]"
        :style="[knobComputedStyle]"
      >
        <div data-testid="resize-bounding-knob" :class="[classNames.knob]">
          <slot v-if="$slots.default"></slot>
        </div>
      </div>
    </div>
  </div>
</template>

<script setup lang="ts">
import {
  onMounted,
  onUnmounted,
  ref,
  type Ref,
  computed,
  type HTMLAttributes,
} from "vue";

import { ResizeBounding } from "../shared/typings.js";

const props = withDefaults(defineProps<Props>(), {
  direction: PaneDirections.RIGHT,
});

const emits = defineEmits<{
  (e: Emits.DRAG_START, data: PaneEmittedData): void;
  (e: Emits.DRAG_MOVE, data: PaneEmittedData): void;
  (e: Emits.DRAG_END, data: PaneEmittedData): void;
  (e: Emits.FOCUS, isFocused: boolean): void;
}>();

const refPane: Ref<HTMLDivElement | null> = ref(null);

const paneComputedStyle = computed(() => {
  const _width = props.options?.width ?? 1;

  if (refPane.value && _width) {
    const _styles = paneBaseStyles(_width);
    const value: HTMLAttributes["style"] =
      _styles[props.direction as PaneDirections];
    return value;
  }

  return;
});

const splitterComputedStyle = computed(() => {
  const _width = props.options?.width;

  if (refPane.value && _width) {
    const _styles = splitterBaseStyles(
      _width,
      props.options?.position ?? ResizeBounding.SplitterPositions.CENTER,
    );
    const value: HTMLAttributes["style"] =
      _styles[props.direction as PaneDirections];
    return value;
  }
});

const isFocused = ref(false);
const isPressed = ref(false);

let isResizing = false;

const knobComputedStyle = computed(() => {
  const isHorizontal = checkIsHorizontal();
  return { transform: `rotate(${isHorizontal ? 90 : 0}deg)` };
});

const checkIsHorizontal = (): boolean =>
  new RegExp(
    `[${PaneDirections.LEFT} | ${PaneDirections.RIGHT} | ${PaneDirectionAliases.HORIZONTAL}]`,
  ).test(props.direction);
const checkIsVertical = (): boolean =>
  new RegExp(
    `[${PaneDirections.TOP} | ${PaneDirections.BOTTOM} | ${PaneDirectionAliases.VERTICAL}]`,
  ).test(props.direction);

const updateCursor = (state: boolean) => {
  let cursorActive: string;

  if (checkIsHorizontal()) {
    cursorActive = props.options?.cursor?.horizontal ?? "auto";
  } else if (checkIsVertical()) {
    cursorActive = props.options?.cursor?.vertical ?? "auto";
  } else cursorActive = "auto";

  if (refPane.value) refPane.value.style.cursor = state ? cursorActive : "auto";
};

const emitFocus = (state: boolean) => {
  emits(Emits.FOCUS, state);
};

const onFocus = (e: PointerEvent, isFocused: boolean): void => {
  e.stopPropagation();
  onSelected(isFocused);
};

const onSelected = (state: boolean): void => {
  isFocused.value = state;
  updateCursor(state);
  emitFocus(state);
};

const onDragStart = (e: PointerEvent): void => {
  e.preventDefault();
  e.stopImmediatePropagation();

  if (!props.options.touchActions && e.pointerType === "touch") return;

  isResizing = true;
  isPressed.value = true;

  onSelected(isResizing);

  const el = e.currentTarget as HTMLDivElement;
  el.setPointerCapture(e.pointerId);

  emits(Emits.DRAG_START, {
    x: Math.round(e.clientX),
    y: Math.round(e.clientY),
    dir: props.direction,
  });

  const onDragMove = (e: PointerEvent): void => {
    emits(Emits.DRAG_MOVE, {
      x: Math.round(e.clientX),
      y: Math.round(e.clientY),
      dir: props.direction,
    });
  };

  const onDragEnd = (e: PointerEvent): void => {
    isResizing = false;
    isPressed.value = false;

    updateCursor(isResizing);

    const el = e.currentTarget as HTMLDivElement;
    el.releasePointerCapture(e.pointerId);

    el.removeEventListener("pointermove", onDragMove);
    el.removeEventListener("pointerup", onDragEnd);

    emits(Emits.DRAG_END, {
      x: Math.round(e.clientX),
      y: Math.round(e.clientY),
      dir: props.direction,
    });
  };

  el.addEventListener("pointermove", onDragMove);
  el.addEventListener("pointerup", onDragEnd);
};

const onDragCancel = (e: PointerEvent): void => {
  const el = e.currentTarget as HTMLDivElement;

  isResizing = false;
  el.releasePointerCapture(e.pointerId);

  emits(Emits.DRAG_END, {
    x: Math.round(e.clientX),
    y: Math.round(e.clientY),
    dir: props.direction,
  });
};

const addEventListeners = () => {
  const el = refPane.value;

  if (el) {
    el.addEventListener("pointerenter", (e) => onFocus(e, true));
    el.addEventListener("pointerleave", (e) => onFocus(e, false));

    el.addEventListener("pointerdown", onDragStart);
    el.addEventListener("pointercancel", onDragCancel);
  }
};

const removeEventListeners = () => {
  const el = refPane.value;

  if (el) {
    el.removeEventListener("pointerenter", (e) => onFocus(e, true));
    el.removeEventListener("pointerleave", (e) => onFocus(e, false));

    el.removeEventListener("pointerdown", onDragStart);
    el.addEventListener("pointercancel", onDragCancel);
  }
};

onMounted(addEventListeners);
onUnmounted(removeEventListeners);

defineExpose({ refPane });
</script>

<script lang="ts">
export interface Props {
  prefix: string;
  direction: PaneDirectionKey;
  options: ResizeBounding.Options;
  classNames: ResizeBounding.IResizeBoundingClassNames;
  styles?: Partial<ResizeBounding.IStyles>;
}

export enum PaneDirectionAliases {
  HORIZONTAL = "h",
  VERTICAL = "v",
}

export enum PaneDirections {
  LEFT = "l",
  RIGHT = "r",
  BOTTOM = "b",
  TOP = "t",
}

export type PaneDirectionKey = PaneDirections | PaneDirections | string;

export enum Emits {
  FOCUS = "focus",
  DRAG_START = "drag:start",
  DRAG_MOVE = "drag:move",
  DRAG_END = "drag:end",
}

export interface PaneEmittedData {
  x: number;
  y: number;
  dir: string;
}

export const paneBaseStyles = (
  size: number,
): Record<PaneDirections, HTMLAttributes["style"]> => {
  const _offset = `${size / 2}px`;

  return {
    l: { top: "0px", left: _offset, width: "0px", height: "100%" },
    r: { top: "0px", right: _offset, width: "0px", height: "100%" },
    t: { left: "0px", top: _offset, width: "100%", height: "0px" },
    b: { left: "0px", bottom: _offset, width: "100%", height: "0px" },
  };
};

export const splitterBaseStyles = (
  size: number,
  position: ResizeBounding.PanePosition,
): Record<PaneDirections, HTMLAttributes["style"]> => {
  const _size = `${size}px`;

  let _offset: string = "0px";

  switch (position) {
    case ResizeBounding.SplitterPositions.CENTER:
      _offset = `0px`;
      break;
    case ResizeBounding.SplitterPositions.EXTERNAL:
      _offset = `${size / 2}px`;
      break;
    case ResizeBounding.SplitterPositions.INTERNAL:
      _offset = `-${size / 2}px`;
      break;
  }

  return {
    l: { right: _offset, width: _size, height: "100%" },
    r: { left: _offset, width: _size, height: "100%" },
    t: { bottom: _offset, height: _size, width: "100%" },
    b: { top: _offset, height: _size, width: "100%" },
  };
};
</script>
