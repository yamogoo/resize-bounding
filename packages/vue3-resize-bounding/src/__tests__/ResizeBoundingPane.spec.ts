import { describe, test, expect, beforeEach, vi } from "vitest";
import { mount, shallowMount, VueWrapper } from "@vue/test-utils";
import { nextTick } from "vue";

import { DataTestIds, DEFAULT_PREFIX } from "./setup";

import deepmerge from "deepmerge";

import ResizeBoundingPane, {
  Emits,
  PaneDirectionAliases,
  PaneDirections,
  paneBaseStyles,
  splitterBaseStyles,
  type PaneEmittedData,
  type Props,
} from "../components/ResizeBoundingPane.vue";
import { ResizeBounding } from "../shared/typings";

import { defaultOptions } from "../components/ResizeBounding.vue";
import { getClassNames } from "../components/ResizeBounding.classNames.js";

const classNames = getClassNames({});

const requiredProps = {
  options: defaultOptions,
  classNames,
};

describe("Boundarize", () => {
  beforeEach(() => {
    Element.prototype.setPointerCapture = vi.fn();
    Element.prototype.releasePointerCapture = vi.fn();
  });

  describe("classes", () => {
    test('should have "normal" class by default', () => {
      const wrapper = shallowMount(ResizeBoundingPane, {
        props: {
          prefix: DEFAULT_PREFIX,
          classNames,
          options: {
            prefix: "",
            width: 0,
            position: "",
            cursor: undefined,
            knob: undefined,
            addStateClasses: true,
            touchActions: false,
          },
        },
      });

      const splitterEl = wrapper.find(`[data-testid="${DataTestIds.PANE}"]`);

      expect(splitterEl.classes("normal")).toBeTruthy();
      expect(splitterEl.classes("normal")).toMatchSnapshot();
    });
  });

  describe("classes", () => {
    test('should have "focused" class', async () => {
      const wrapper = shallowMount(ResizeBoundingPane, {
        props: {
          prefix: DEFAULT_PREFIX,
          classNames,
          options: {
            prefix: "",
            width: 0,
            position: "",
            cursor: undefined,
            knob: undefined,
            addStateClasses: true,
            touchActions: false,
          },
        },
      });

      const paneEl = wrapper.find(`[data-testid="${DataTestIds.PANE}"]`);

      await paneEl.trigger("pointerenter");

      expect(paneEl.classes("focused")).toBeTruthy();
      expect(paneEl.classes("focused")).toMatchSnapshot();
    });
  });

  describe("events", () => {
    describe("focus", async () => {
      const wrapper = shallowMount(ResizeBoundingPane, {
        props: {
          prefix: DEFAULT_PREFIX,
          direction: PaneDirections.RIGHT,
          ...requiredProps,
        },
      });

      const rootEl = wrapper.find(`[data-testid="${DataTestIds.PANE}"]`);

      test(`should emit "${Emits.FOCUS}" with state (true)`, async () => {
        await rootEl.trigger("pointerenter");
        const isFocused: boolean = (
          wrapper.emitted(Emits.FOCUS) as boolean[][]
        )[0][0];

        expect(isFocused).toBeTruthy();
        expect(isFocused).toMatchSnapshot();
      });

      test(`should emit "${Emits.FOCUS}" with state (false)`, async () => {
        await rootEl.trigger("pointerleave");
        const isFocused = (wrapper.emitted(Emits.FOCUS) as boolean[][])[1][0];

        expect(isFocused).toBeFalsy();
        expect(isFocused).toMatchSnapshot();
      });
    });

    describe("drag", () => {
      const wrapper = shallowMount(ResizeBoundingPane, {
        props: {
          prefix: DEFAULT_PREFIX,
          direction: PaneDirections.RIGHT,
          ...requiredProps,
        },
      });

      const rootEl = wrapper.find(`[data-testid="${DataTestIds.SPLITTER}"]`);

      test(`should emit "${Emits.DRAG_START} -> ${Emits.DRAG_MOVE} -> ${Emits.DRAG_END}" with {x, y, dir}`, async () => {
        const checkData = (
          eventName: string,
          wrapper: VueWrapper,
          eventData: { clientX: number; clientY: number },
        ): PaneEmittedData => {
          const ev = wrapper.emitted(eventName) as PaneEmittedData[][];
          const data = ev[0][0];
          const { x, y } = data;

          const { clientX, clientY } = eventData;

          expect(data).toHaveProperty("x");
          expect(x).toBe(clientX);
          expect(data).toHaveProperty("y");
          expect(y).toBe(clientY);
          expect(data).toHaveProperty("dir");
          expect(data.dir).toBe(PaneDirections.RIGHT);
          return data;
        };

        let events: PaneEmittedData[] = [];

        const makeTrigger = async (
          event: string,
          eventData: { clientX: number; clientY: number },
          emittedEventName: string,
        ) => {
          await rootEl.trigger(event, eventData);
          events.push(checkData(emittedEventName, wrapper, eventData));
        };

        await rootEl.trigger("pointerenter", { clientX: 1, clientY: 1 });
        await makeTrigger(
          "pointerdown",
          { clientX: 1, clientY: 1 },
          Emits.DRAG_START,
        );
        await makeTrigger(
          "pointermove",
          { clientX: 2, clientY: 2 },
          Emits.DRAG_MOVE,
        );
        await makeTrigger(
          "pointerup",
          { clientX: 3, clientY: 3 },
          Emits.DRAG_END,
        );

        expect(events.length).toBe(3);
        expect(events).toMatchSnapshot();
      });
    });
  });

  describe("options", () => {
    describe("knob", () => {
      test("should always show knob", () => {
        const wrapper = shallowMount(ResizeBoundingPane, {
          props: {
            prefix: DEFAULT_PREFIX,
            classNames,
            styles: {},
            options: deepmerge(defaultOptions, {
              knob: {
                show: true,
                constantlyShow: true,
              },
            }),
          },
        });

        const knobEl = wrapper.find(`[data-testid="${DataTestIds.KNOB}"]`);

        expect(knobEl.exists()).toBeTruthy();
        expect(knobEl.attributes("data-testid")).toMatchSnapshot();
      });

      test.each([
        [PaneDirections.LEFT, 90],
        [PaneDirections.RIGHT, 90],
        [PaneDirections.TOP, 0],
        [PaneDirections.BOTTOM, 0],
      ])("pane (--%s) should rotated to (%s deg)", (direction, deg) => {
        const wrapper = shallowMount(ResizeBoundingPane, {
          props: {
            prefix: DEFAULT_PREFIX,
            direction,
            styles: {},
            classNames,
            options: deepmerge(defaultOptions, {
              knob: {
                show: true,
                constantlyShow: true,
              },
            }),
          },
        });

        const knobEl = wrapper.find(
          `[data-testid="${DataTestIds.SPLITTER_CONTAINER}"]`,
        );
        const styles = knobEl.attributes("style");

        expect(styles).toContain(`(${deg}deg)`);
        expect(styles).toMatchSnapshot();
      });
    });
  });

  describe("styles", () => {
    describe("cursor", () => {
      test('default cursor style should be "auto"', async () => {
        const wrapper = shallowMount(ResizeBoundingPane, {
          props: {
            prefix: DEFAULT_PREFIX,
            ...requiredProps,
          },
        });
        await wrapper.setProps({
          direction: PaneDirections.RIGHT,
          prefix: DEFAULT_PREFIX,
        });

        const rootEl = wrapper.find(`[data-testid="${DataTestIds.PANE}"]`);

        await rootEl.trigger("pointerenter");
        await rootEl.trigger("pointerleave");

        const cursorStyle = rootEl.attributes("style");

        expect(cursorStyle).toContain("auto");
        expect(cursorStyle).toMatchSnapshot();
      });

      const checkCursorActive = async (
        wrapper: VueWrapper,
        props: Partial<Props>,
        comparedCursorStyle: string,
      ): Promise<void> => {
        await wrapper.setProps(props);

        const rootEl = wrapper.find(`[data-testid="${DataTestIds.PANE}"]`);
        await rootEl.trigger("pointerenter");

        const cursorStyle = rootEl.attributes("style");

        expect(cursorStyle).toContain(comparedCursorStyle);
        expect(cursorStyle).toMatchSnapshot();
      };

      describe("active", () => {
        describe("default styles", () => {
          test.each([
            PaneDirectionAliases.HORIZONTAL,
            PaneDirections.RIGHT,
            PaneDirections.LEFT,
          ])(
            'default cursor style should be "col-resize" (--%s)',
            async (direction) => {
              const wrapper = shallowMount(ResizeBoundingPane, {
                props: {
                  prefix: DEFAULT_PREFIX,
                  ...requiredProps,
                },
              });

              checkCursorActive(wrapper, { direction }, "col-resize");
            },
          );

          test.each([
            PaneDirectionAliases.VERTICAL,
            PaneDirections.TOP,
            PaneDirections.BOTTOM,
          ])(
            'default cursor style should be "row-resize" (--%s)',
            async (direction) => {
              const wrapper = shallowMount(ResizeBoundingPane, {
                props: {
                  prefix: DEFAULT_PREFIX,
                  ...requiredProps,
                },
              });

              checkCursorActive(wrapper, { direction }, "row-resize");
            },
          );
        });

        describe("custom styles (options)", () => {
          const CUSTOM_ACTIVE_CURSOR = "ew-resize";

          test.each([...Object.values(PaneDirections)])(
            `custom cursor style should be "${CUSTOM_ACTIVE_CURSOR}" (--%s)`,
            async (direction) => {
              const wrapper = shallowMount(ResizeBoundingPane, {
                props: {
                  prefix: DEFAULT_PREFIX,
                  ...requiredProps,
                },
              });

              checkCursorActive(
                wrapper,
                {
                  direction,
                  options: {
                    cursor: {
                      vertical: CUSTOM_ACTIVE_CURSOR,
                      horizontal: CUSTOM_ACTIVE_CURSOR,
                    },
                    width: 0,
                    position: "central",
                    knob: undefined,
                    prefix: "",
                    addStateClasses: false,
                    touchActions: false,
                  },
                },
                CUSTOM_ACTIVE_CURSOR,
              );
            },
          );

          const SIZE = 12;
          const paneMap = paneBaseStyles(
            SIZE,
            SIZE,
            ResizeBounding.SplitterPositions.CENTER,
          );
          const splitterMap = splitterBaseStyles(SIZE, SIZE);

          test.each([
            [PaneDirections.LEFT, paneMap.l],
            [PaneDirections.RIGHT, paneMap.r],
            [PaneDirections.TOP, paneMap.t],
            [PaneDirections.BOTTOM, paneMap.b],
          ])(
            `the pane (--%s) should have "%s" styles`,
            async (direction, styles) => {
              const wrapper = shallowMount(ResizeBoundingPane, {
                props: {
                  prefix: DEFAULT_PREFIX,
                  direction,
                  classNames,
                  options: { ...defaultOptions, width: SIZE },
                },
              });

              await nextTick();

              const paneEl = wrapper.find(
                `[data-testid="${DataTestIds.PANE}"]`,
              );
              const style = paneEl.attributes("style");

              if (styles) {
                for (const [k, v] of Object.entries(styles)) {
                  expect(style).toContain(`${k}: ${v}`);
                }

                expect(style).toMatchSnapshot();
              }
            },
          );

          test.each([
            [PaneDirections.LEFT, splitterMap.l],
            [PaneDirections.RIGHT, splitterMap.r],
            [PaneDirections.TOP, splitterMap.t],
            [PaneDirections.BOTTOM, splitterMap.b],
          ])(
            `the pane splitter (--%s) should have "%s" styles`,
            async (direction, styles) => {
              const wrapper = shallowMount(ResizeBoundingPane, {
                props: {
                  prefix: DEFAULT_PREFIX,
                  direction,
                  classNames,
                  options: { ...defaultOptions, width: SIZE },
                },
              });

              await nextTick();

              const splitterEl = wrapper.find(
                `[data-testid="${DataTestIds.SPLITTER}"]`,
              );
              const style = splitterEl.attributes("style");

              if (styles) {
                for (const [k, v] of Object.entries(styles)) {
                  expect(style).toContain(`${k}: ${v}`);
                }

                expect(style).toMatchSnapshot();
              }
            },
          );
        });
      });
    });

    describe("custom styles", () => {
      test.each([
        [{ left: "70px", right: "70px", width: "70px" }, PaneDirections.RIGHT],
      ])(
        "inner styles of splitter should not have to overrided by customStyles",
        async (_customStyles, direction) => {
          const SPLITTER_WIDTH = 12;

          const styles = splitterBaseStyles(SPLITTER_WIDTH, SPLITTER_WIDTH)[
            direction
          ];

          const wrapper = shallowMount(ResizeBoundingPane, {
            props: {
              prefix: DEFAULT_PREFIX,
              direction,
              classNames,
              options: {
                ...defaultOptions,
                width: SPLITTER_WIDTH,
              },
            },
          });

          const splitterEl = wrapper.find(
            `[data-testid="${DataTestIds.SPLITTER}"]`,
          );

          await splitterEl.trigger("pointerenter");
          const splitterStyles = splitterEl.attributes("style");

          if (styles && splitterStyles) {
            for (const [k, v] of Object.entries(styles)) {
              const value = `${k}: ${v.toString()}`.trim();
              const trimedStyles = splitterStyles.toString().trim();

              expect(trimedStyles).toContain(value);
              expect(value).toMatchSnapshot();
              expect(trimedStyles).toMatchSnapshot();
            }
            expect(splitterStyles).toMatchSnapshot();
          }
        },
      );
    });
  });

  describe("slots", () => {
    test.each(["<p>knob</p>"])("should render knob slot (%s)>", (knobSlot) => {
      const wrapper = mount(ResizeBoundingPane, {
        props: {
          prefix: DEFAULT_PREFIX,
          direction: "r",
          classNames,
          options: {
            knob: {
              show: true,
              normalHidden: false,
            },
            width: 3,
            position: "central",
            cursor: {},
            prefix: "",
            addStateClasses: true,
            touchActions: false,
          },
        },
        slots: {
          default: knobSlot,
        },
      });

      const text = wrapper.find(`[data-testid="${DataTestIds.KNOB}"]`).text();
      expect(text).toBe("knob");
      expect(text).toMatchSnapshot();
    });
  });
});
