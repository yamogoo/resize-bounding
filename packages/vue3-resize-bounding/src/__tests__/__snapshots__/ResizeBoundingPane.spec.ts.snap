// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Boundarize > classes > should have "focused" class 1`] = `true`;

exports[`Boundarize > classes > should have "normal" class by default 1`] = `true`;

exports[`Boundarize > events > drag > should emit "drag:start -> drag:move -> drag:end" with {x, y, dir} 1`] = `
[
  {
    "dir": "r",
    "x": 1,
    "y": 1,
  },
  {
    "dir": "r",
    "x": 2,
    "y": 2,
  },
  {
    "dir": "r",
    "x": 3,
    "y": 3,
  },
]
`;

exports[`Boundarize > events > focus > should emit "focus" with state (false) 1`] = `false`;

exports[`Boundarize > events > focus > should emit "focus" with state (true) 1`] = `true`;

exports[`Boundarize > options > knob > pane (--b) should rotated to (0 deg) 1`] = `"transform: rotate(0deg);"`;

exports[`Boundarize > options > knob > pane (--l) should rotated to (90 deg) 1`] = `"transform: rotate(90deg);"`;

exports[`Boundarize > options > knob > pane (--r) should rotated to (90 deg) 1`] = `"transform: rotate(90deg);"`;

exports[`Boundarize > options > knob > pane (--t) should rotated to (0 deg) 1`] = `"transform: rotate(0deg);"`;

exports[`Boundarize > options > knob > should always show knob 1`] = `"resize-bounding-knob"`;

exports[`Boundarize > slots > should render knob slot (<p>knob</p>)> 1`] = `"knob"`;

exports[`Boundarize > styles > cursor > active > custom styles (options) > custom cursor style should be "ew-resize" (--b) 1`] = `"cursor: ew-resize;"`;

exports[`Boundarize > styles > cursor > active > custom styles (options) > custom cursor style should be "ew-resize" (--l) 1`] = `"cursor: ew-resize;"`;

exports[`Boundarize > styles > cursor > active > custom styles (options) > custom cursor style should be "ew-resize" (--r) 1`] = `"cursor: ew-resize;"`;

exports[`Boundarize > styles > cursor > active > custom styles (options) > custom cursor style should be "ew-resize" (--t) 1`] = `"cursor: ew-resize;"`;

exports[`Boundarize > styles > cursor > active > custom styles (options) > the pane (--b) should have "{ left: '0px', bottom: '-6px', width: '100%', height: '12px' }" styles 1`] = `"left: 0px; bottom: -6px; width: 100%; height: 12px;"`;

exports[`Boundarize > styles > cursor > active > custom styles (options) > the pane (--l) should have "{ top: '0px', left: '-6px', width: '12px', height: '100%' }" styles 1`] = `"top: 0px; left: -6px; width: 12px; height: 100%;"`;

exports[`Boundarize > styles > cursor > active > custom styles (options) > the pane (--r) should have "{ top: '0px', right: '-6px', width: '12px', height: '100%' }" styles 1`] = `"top: 0px; right: -6px; width: 12px; height: 100%;"`;

exports[`Boundarize > styles > cursor > active > custom styles (options) > the pane (--t) should have "{ left: '0px', top: '-6px', width: '100%', height: '12px' }" styles 1`] = `"left: 0px; top: -6px; width: 100%; height: 12px;"`;

exports[`Boundarize > styles > cursor > active > custom styles (options) > the pane splitter (--b) should have "{ top: '0px', height: '12px', width: '100%' }" styles 1`] = `"top: 0px; height: 12px; width: 100%;"`;

exports[`Boundarize > styles > cursor > active > custom styles (options) > the pane splitter (--l) should have "{ right: '0px', width: '12px', height: '100%' }" styles 1`] = `"right: 0px; width: 12px; height: 100%;"`;

exports[`Boundarize > styles > cursor > active > custom styles (options) > the pane splitter (--r) should have "{ left: '0px', width: '12px', height: '100%' }" styles 1`] = `"left: 0px; width: 12px; height: 100%;"`;

exports[`Boundarize > styles > cursor > active > custom styles (options) > the pane splitter (--t) should have "{ bottom: '0px', height: '12px', width: '100%' }" styles 1`] = `"bottom: 0px; height: 12px; width: 100%;"`;

exports[`Boundarize > styles > cursor > active > default styles > default cursor style should be "col-resize" (--h) 1`] = `"cursor: col-resize;"`;

exports[`Boundarize > styles > cursor > active > default styles > default cursor style should be "col-resize" (--l) 1`] = `"top: 0px; left: -2px; width: 4px; height: 100%; cursor: col-resize;"`;

exports[`Boundarize > styles > cursor > active > default styles > default cursor style should be "col-resize" (--r) 1`] = `"top: 0px; right: -2px; width: 4px; height: 100%; cursor: col-resize;"`;

exports[`Boundarize > styles > cursor > active > default styles > default cursor style should be "row-resize" (--b) 1`] = `"left: 0px; bottom: -2px; width: 100%; height: 4px; cursor: row-resize;"`;

exports[`Boundarize > styles > cursor > active > default styles > default cursor style should be "row-resize" (--t) 1`] = `"left: 0px; top: -2px; width: 100%; height: 4px; cursor: row-resize;"`;

exports[`Boundarize > styles > cursor > active > default styles > default cursor style should be "row-resize" (--v) 1`] = `"cursor: row-resize;"`;

exports[`Boundarize > styles > cursor > default cursor style should be "auto" 1`] = `"top: 0px; right: -2px; width: 4px; height: 100%; cursor: auto;"`;

exports[`Boundarize > styles > custom styles > inner styles of splitter should not have to overrided by customStyles 1`] = `"left: 0px"`;

exports[`Boundarize > styles > custom styles > inner styles of splitter should not have to overrided by customStyles 2`] = `"left: 0px; width: 12px; height: 100%;"`;

exports[`Boundarize > styles > custom styles > inner styles of splitter should not have to overrided by customStyles 3`] = `"width: 12px"`;

exports[`Boundarize > styles > custom styles > inner styles of splitter should not have to overrided by customStyles 4`] = `"left: 0px; width: 12px; height: 100%;"`;

exports[`Boundarize > styles > custom styles > inner styles of splitter should not have to overrided by customStyles 5`] = `"height: 100%"`;

exports[`Boundarize > styles > custom styles > inner styles of splitter should not have to overrided by customStyles 6`] = `"left: 0px; width: 12px; height: 100%;"`;

exports[`Boundarize > styles > custom styles > inner styles of splitter should not have to overrided by customStyles 7`] = `"left: 0px; width: 12px; height: 100%;"`;
